### PR TITLE
Optimizations to the Scouting Data Formats

### PR DESCRIPTION
**Pull request by dsperka**
#### PR description:

This PR makes some optimizations to the scouting data format. We store the track quantities for pfCandidates as relative to the pfCandidate itself, which results in a better compression. We also dropped pfCandidate mass, because it is redundant w.r.t. pdgId. We also only store non-zero rechit energies, resulting in smaller data size when also storing the associated detIds. 

To make the changes apparent in a backward compatible way we added new members to the data formats to flag these changes. 

These changes, together with other updates to the HLT configuration not in this PR, result in a scouting event size of ~7.7 kb.

### CODE REVIEW

**Comment by missirol** [comment]
@dsperka 

Please fix the issue with `code-checks`, and update the title of the PR.

File: DataFormats/Scouting/src/classes_def.xml [hunk 1]
```diff
@@ -33,22 +33,27 @@
   <class name="Run3ScoutingPFJet" ClassVersion="3">
       <version ClassVersion="3" checksum="3819751468"/>
   </class>
-  <class name="Run3ScoutingParticle" ClassVersion="4">
+  <class name="Run3ScoutingParticle" ClassVersion="7">
+   <version ClassVersion="7" checksum="2425016900"/>
+   <version ClassVersion="6" checksum="3844610129"/>
+   <version ClassVersion="5" checksum="583559071"/>
```
Comment (makortel) [inline DataFormats/Scouting/src/classes_def.xml:39]: Please drop versions 5 and 6, rename 7 as 5, move it below 4, and use the same indentation.

File: DataFormats/Scouting/src/classes_def.xml [hunk 2]
```diff
@@ -33,22 +33,27 @@
...
       <version ClassVersion="3" checksum="1672617274"/>
       <version ClassVersion="4" checksum="1541586682"/>
   </class>
   <class name="Run3ScoutingVertex" ClassVersion="3">
       <version ClassVersion="3" checksum="316446070"/>
   </class>
-  <class name="Run3ScoutingElectron" ClassVersion="4">
+  <class name="Run3ScoutingElectron" ClassVersion="5">
+   <version ClassVersion="5" checksum="2230390721"/>
```
Comment (makortel) [inline DataFormats/Scouting/src/classes_def.xml:47]: Please move below 4 and use consistent indentation.

File: DataFormats/Scouting/src/classes_def.xml [hunk 3]
```diff
@@ -33,22 +33,27 @@
...
       <version ClassVersion="3" checksum="1086011373"/>
       <version ClassVersion="4" checksum="1250202632"/>
   </class>
   <class name="Run3ScoutingMuon" ClassVersion="4">
       <version ClassVersion="3" checksum="3882497397"/>
       <version ClassVersion="4" checksum="4206297195"/>
   </class>
-  <class name="Run3ScoutingPhoton" ClassVersion="4">
+  <class name="Run3ScoutingPhoton" ClassVersion="5">
+   <version ClassVersion="5" checksum="1238010495"/>
```
Comment (makortel) [inline DataFormats/Scouting/src/classes_def.xml:56]: Please move below 4 and use consistent indentation.

**Review by makortel — COMMENTED** [review]
> To make the changes apparent in a backward compatible way we added new members to the data formats to flag these changes.

Technically already adding members breaks the reading of old streamer files (and for ROOT more complex evolution would work as well).

Has any data been written in streamer files already with the earlier version?

File: HLTrigger/Egamma/plugins/HLTScoutingEgammaProducer.cc [hunk 4]
```diff
@@ -212,16 +213,23 @@ void HLTScoutingEgammaProducer::produce(edm::StreamID sid, edm::Event& iEvent, e
...
     for (int i = 0; i < detSize; i++) {
-      mDetIdIds[i] = mDetIds[i];
-      mEnergies[i] =
-          MiniFloatConverter::reduceMantissaToNbitsRounding(recHitE(mDetIds[i], *rechits), mantissaPrecision);
-      if (saveRecHitTiming)
-        mTimes[i] = MiniFloatConverter::reduceMantissaToNbitsRounding(recHitT(mDetIds[i], *rechits), mantissaPrecision);
+      if (not rechitZeroSuppression or recHitE(mDetIds[i], *rechits) > 0) {
+        mDetIdIds.push_back(mDetIds[i]);
+        mEnergies.push_back(
+            MiniFloatConverter::reduceMantissaToNbitsRounding(recHitE(mDetIds[i], *rechits), mantissaPrecision));
```
Comment (missirol) [inline HLTrigger/Egamma/plugins/HLTScoutingEgammaProducer.cc:227]:
```suggestion
      auto const recHit_en = recHitE(mDetIds[i], *rechits);
      if (not rechitZeroSuppression or recHit_en > 0) {
        mDetIdIds.push_back(mDetIds[i]);
        mEnergies.push_back(
            MiniFloatConverter::reduceMantissaToNbitsRounding(recHit_en, mantissaPrecision));
```
If `rechitZeroSuppression==false`, `recHitE` is calculated once (L227). If `rechitZeroSuppression==true`, `recHitE` is calculated once (L224) or twice (L224 and L227). This should avoid calculating it twice.

**Comment by dsperka** [comment]
I don't know if streamer files have been written with previous versions. If yes and thats a problem, we need some mechanism to evolve these data formats. It is not reasonable to assume there will be no changes through the next 4 years.

**Comment by fwyzard** [comment]
> > To make the changes apparent in a backward compatible way we added new members to the data formats to flag these changes.
> 
> Technically already adding members breaks the reading of old streamer files (and for ROOT more complex evolution would work as well).
> 
> Has any data been written in streamer files already with the earlier version?

@makortel , why do we care about the streamer files ?
they are supposed to be very transient, living only for a couple of days from the data taking at HLT to the repacking at Tier-0, and certainly don't need to be compatible across releases (_i.e._ CMSSW 12.4.x vs 12.3.x).

**Comment by makortel** [comment]
> why do we care about the streamer files ?
> they are supposed to be very transient, living only for a couple of days from the data taking at HLT to the repacking at Tier-0, and certainly don't need to be compatible across releases (_i.e._ CMSSW 12.4.x vs 12.3.x).

Tier0 uses streamer files for replays. I understood from @drkovalskyi that if really needed, scouting streams could in principle be ignored in Tier0 tests, but that would require manual work (and tis thus preferred to be avoided).

Another consideration is if there is any chance for a situation where 12_3_X would be used in HLT, and 12_4_X in Tier0 (even for a short amount of time).

If scouting streams have not been used yet in the commissioning data taking, and if they would not be enabled before the deployment of 12_4_X in HLT, I'd think this update would be safe enough.

@dsperka You mentioned in the description that with this PR (and updates in HLT configuration) the scouting event size would be ~7.7 kB. What would the event size be without this PR? Just to understand the importance for data volume.

**Comment by dsperka** [comment]
Without these changes the size would be ~10.9 kB. 

I would like to repeat though that we need to have a way to evolve the data formats. We want to improve the scouting data throughout Run 3, and if we cannot ever change the data format that is extremely difficult.

**Comment by fwyzard** [comment]
> Tier0 uses streamer files for replays. I understood from @drkovalskyi that if really needed, scouting streams could in principle be ignored in Tier0 tests, but that would require manual work (and tis thus preferred to be avoided).
>
> Another consideration is if there is any chance for a situation where 12_3_X would be used in HLT, and 12_4_X in Tier0 (even for a short amount of time).


The streamer format has _never_ been required or guaranteed to be readable with _any_ version other than the _exact same one_ with which it has been written. Any compatibility across minor or major releases is accidental, and any workflows that (ab)uses it has been working (or not) by chance.

Whenever the release used for prompt reco and HLT have diverged in the past, the solution has been to keep the release used for the repacking step in sync with HLT, and allow a different (more recent) release for the prompt reco step, based on the guaranteed forward compatibility of the RAW data written in .root format.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
Looks like you need to code (and call) a customisation function inside https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
which removes the obsolete ```doTrackRelVars``` when present in any instance of the plugin. See the file for examples.

**Comment by makortel** [comment]
@cmsbuild, please test

For cleaner comparisons

**Comment by makortel** [comment]
+core

* ~30 % reduction in scouting event size
* I was overconcerned earlier on the impact to Tier0

**Comment by missirol** [comment]
+hlt

**Comment by perrotta** [comment]
+1

**Pull request merged**